File: WdtcCore/src/main/java/org/wdt/wdtc/download/forge/ForgeVersionList.java
package org.wdt.wdtc.download.forge;


import org.wdt.utils.gson.JSONArray;
import org.wdt.utils.gson.JSONObject;
import org.wdt.wdtc.download.infterface.VersionList;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.utils.PlatformUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ForgeVersionList implements VersionList {
    private final Launcher launcher;


    public ForgeVersionList(Launcher launcher) {
        this.launcher = launcher;
    }

    public String getForgeListUrl() {
        return "https://bmclapi2.bangbang93.com/forge/minecraft/" + launcher.getVersionNumber();
    }

    @Override
    public List<String> getVersionList() throws IOException {
        List<String> VersionName = new ArrayList<>();
        JSONArray VersionList = JSONArray.parseJSONArray(PlatformUtils.UrltoString(getForgeListUrl()));
        for (int i = 0; i < VersionList.size(); i++) {
            JSONObject VersionObject = VersionList.getJSONObject(i);
            VersionName.add(VersionObject.getString("version"));
        }
        return VersionName;
    }
}

File: core/src/main/kotlin/org/wdt/wdtc/core/impl/download/forge/ForgeVersionList.kt
package org.wdt.wdtc.core.impl.download.forge

import com.google.gson.annotations.SerializedName
import org.wdt.utils.gson.parseJsonArray
import org.wdt.utils.io.toStrings
import org.wdt.wdtc.core.openapi.download.interfaces.VersionListInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsJsonObjectInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence.Companion.asSequence
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence.Companion.asVersionsObjectSequence
import org.wdt.wdtc.core.openapi.game.Version
import org.wdt.wdtc.core.openapi.utils.ioAsync
import org.wdt.wdtc.core.openapi.utils.toURL
import java.util.*

class ForgeVersionList(version: Version) : VersionListInterface {
	private val forgeListUrl = "https://bmclapi2.bangbang93.com/forge/minecraft/${version.versionNumber}"
	
	private val list = ioAsync { forgeListUrl.toURL().toStrings().parseJsonArray() }
	
	override suspend fun getVersionList(): VersionsObjectSequence {
		return list.await().asSequence<ForgeVersionsJsonObjectImpl>().asVersionsObjectSequence()
	}
	
	
	class ForgeVersionsJsonObjectImpl(
		@field:SerializedName("version") override val versionNumber: String,
		
		@field:SerializedName("modified") val modified: Date,
	) : VersionsJsonObjectInterface {
		
		@SerializedName("mcversion")
		val mcversion: String? = null
		override fun equals(other: Any?): Boolean {
			if (this === other) return true
			if (other == null || javaClass != other.javaClass) return false
			val that = other as ForgeVersionsJsonObjectImpl
			return versionNumber == that.versionNumber && mcversion == that.mcversion
		}
		
		override fun hashCode(): Int {
			return Objects.hash(versionNumber, mcversion)
		}
	}
	
}
